tensorcircuit/templates/ensemble.py

```diff
@@ -0,0 +1,165 @@
+"""
+Useful utilities for ensemble
+"""
+
+import tensorflow as tf
+import numpy as np
+
+
+class bagging:  # A.K.A. voting
+    def __init__(self):
+        self.models = []
+        self.is_trained = []
+        self.count = 0
+        self.need_confidence = True  # Help in reducing numbers of get_confidence runs
+        self.permit_train = False
+
+    def append(self, model, is_trained):
+        """
+        Add model to the voting method
+        """
+        self.models.append(model)
+        self.is_trained.append(is_trained)
+        self.count += 1
+
+    def __train_model(self, i, **kwargs):
+        """
+        Train a model if it isn't trained already
+        """
+        if not self.is_trained[i]:
+            self.need_confidence = True
+            self.is_trained[i] = True
+            self.models[i]
+            self.models[i].trainable
+            self.models[i].fit(**kwargs)
+
+    def train(self, **kwargs):
+        """
+        Train all models in the class, **kwargs expect to receive the argus that can be directly sent to tf.fit
+        Expected to be run after finishing compile
+        """
+        if not self.permit_train:
+            raise Exception("Needed to be compiled before training")
+        for i in range(self.count):
+            if "verbose" in kwargs:
+                if kwargs["verbose"] == 1:
+                    print("Model ", i + 1, "/", self.count, " is training...")
+            else:
+                print("Model ", i + 1, "/", self.count, " is training...")
+            self.__train_model(i, **kwargs)
+
+    def compile(self, **kwargs):
+        """
+        Compile code
+        """
+        self.permit_train = True
+        for i in range(self.count):
+            self.models[i].compile(**kwargs)
+
+    def __get_confidence(self, model_index, input):
+        """
+        Get the confidence value that is needed by voting.
+        Number of calling this function is reduced by self.need_confidence
+        """
+        self.need_confidence = False
+        prediction = self.models[model_index].predict(input, verbose=0)
+        prediction_returns = np.zeros(len(prediction))
+        for i in range(len(prediction)):
+            prediction_returns[i] = prediction[i][0]
+        return prediction_returns
+
+    """
+    Voting strategies begin
+    More voting strategies can be added beneath, a single function, and a if function in self.predict
+    """
+
+    def __voting_weight(self, array):
+        result = []
+        for i in array:
+            result.append(self.__voting_weight_single(i))
+        return result
+
+    def __voting_most(self, array):
+        result = []
```

Comment:
we can actually do the voting in batch form instead of a for and two functions (with and without _single).

For example, we have ``np.mean(array, axis=1)`` where the axis can be tuned